src/ansiblelint/transformer.py

```diff
@@ -0,0 +1,164 @@
+# So, we need to identify whitespace-only lines to drop spaces before reading.
+_whitespace_only_lines_re = re.compile(r"^ +$", re.MULTILINE)
+
+
+class FormattedEmitter(Emitter):
+    """Emitter that applies custom formatting rules when dumping YAML.
+
+    Root-level sequences are never indented.
+    All subsequent levels are indented as configured (normal ruamel.yaml behavior).
+
+    Earlier implementations used dedent on ruamel.yaml's dumped output,
+    but string magic like that had a ton of problematic edge cases.
+    """
+
+    _sequence_indent = 2
+    _sequence_dash_offset = 0  # Should be _sequence_indent - 2
+
+    # NB: mypy does not support overriding attributes with properties yet:
+    #     https://github.com/python/mypy/issues/4125
+    #     To silence we have to ignore[override] both the @property and the method.
+
+    @property  # type: ignore[override]
+    def best_sequence_indent(self) -> int:  # type: ignore[override]
+        """Return the configured sequence_indent or 2 for root level."""
+        return 2 if self.column < 2 else self._sequence_indent
+
+    @best_sequence_indent.setter
+    def best_sequence_indent(self, value: int) -> None:
+        """Configure how many columns to indent each sequence item (including the '-')."""
+        self._sequence_indent = value
+
+    @property  # type: ignore[override]
+    def sequence_dash_offset(self) -> int:  # type: ignore[override]
+        """Return the configured sequence_dash_offset or 2 for root level."""
+        return 0 if self.column < 2 else self._sequence_dash_offset
+
+    @sequence_dash_offset.setter
+    def sequence_dash_offset(self, value: int) -> None:
+        """Configure how many spaces to put before each sequence item's '-'."""
+        self._sequence_dash_offset = value
+
+
+# Transformer is for transforms like runner is for rules
+class Transformer:
+    """Transformer class performs the fmt transformations."""
+
+    def __init__(self, result: LintResult):
+        """Initialize a Transformer instance."""
+        # TODO: options for explict_start, indent_sequences
+        self.explicit_start = True
+        self.indent_sequences = True
+
+        self.matches: List[MatchError] = result.matches
+        self.files: Set[Lintable] = result.files
+
+        file: Lintable
+        lintables: Dict[str, Lintable] = {file.filename: file for file in result.files}
+        self.matches_per_file: Dict[Lintable, List[MatchError]] = {
+            file: [] for file in result.files
+        }
+
+        for match in self.matches:
+            try:
+                lintable = lintables[match.filename]
+            except KeyError:
+                # we shouldn't get here, but this is easy to recover from so do that.
+                lintable = Lintable(match.filename)
+                self.matches_per_file[lintable] = []
+            self.matches_per_file[lintable].append(match)
+
+    def run(self) -> None:
```

Comment:
I think this function does several things:
- sets up the transformer
- loads the file (call to `load_data`)
- checks if we should operate on the file
- runs the transform (first half of the call to `yaml.dump`)
- writes the file to disk (second half of the call to `yaml.dump`)

Two of these tasks are IO/side-effecting. I think this makes it harder to test. For example, if we want to test that it still works with like 1000 lines of YAML, we have to have a file with 1000 lines of YAML. We couldn't just create it in place, with something like `['expr']*1000`.

I think bundling these together will also lead to duplication between transformers. I suspect that all transformers will need to read and write the file, so it would be worth perhaps extracting to a template method. I think that would also make it easier to create new transformer, by simply needing to subclass transformer and override a `def transform(file: TextIO)->TextIO`

I think we should address this now (as opposed to once this is merged) to make testing easier and so the initial interface is easier to use.

Reply:
Ah, yes, this makes sense. I was thinking that the design should look like that, and it does!
I would still suggest that we extract the actual reformatting of the file (and all that ruamel configuration) into a separate function so it's easier to test. It would also be easier to implement it as a TransformMixin.transform later

Reply:
> I would still suggest that we extract the actual reformatting of the file (and all that ruamel configuration) into a separate function so it's easier to test. It would also be easier to implement it as a TransformMixin.transform later

OK. So, I pulled out the ruamel.yaml config, but I'm not clear on how I might reorganize the `run()` method for testability. Could you take a look at the latest code and see if you see a good way to break it up?

Reply:
I think the key thing would be separating the "formatting yaml" part from the "reading/writing files". I think we could just separate the `yaml.dump` into `yaml.dumps` and `file.write`, and then scoot the `yaml.write` into a function like 
```python
def reformat(self, unformatted):
    return self.yaml.dumps(unformatted)
```
Then we can have a few integration tests for actually reading and writing files, and then many more unit tests for covering weird YAML things.

We could get really into breaking up the other things (separating the is_yaml filtering), but I think the above is a good step.

Reply:
Ah. The joys of `ruamel.yaml`. It does not provide a `dumps`. It will only write to a stream pushing people to let it write the file directly. So, to get around that I would have to create a `StringIO` and then handle writing it myself.
I'm doing that for the test I'm writing, but it's awkward.